File: DailyLeetcode/28Aug/SortMatrixbyDiagonal.java
//Time Complexity: O(m * n * log(min(m, n))) where m is the number of rows and n is the number of columns in the matrix.
//Space Complexity: O(m * n) for storing the elements in the priority queues.

import java.util.*;

class Solution {
    public int[][] sortMatrix(int[][] matrix) {
        Map<Integer, PriorityQueue<Integer>> diagonalMap = new HashMap<>();
        int rows = matrix.length, cols = matrix[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int key = i - j;
                diagonalMap.putIfAbsent(key, key < 0 ? new PriorityQueue<>() 
                                                     : new PriorityQueue<>(Collections.reverseOrder()));
                diagonalMap.get(key).offer(matrix[i][j]);
            }
        }

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                int key = i - j;
                matrix[i][j] = diagonalMap.get(key).poll();
            }
        }

        return matrix;
    }
}

File: DailyLeetcode/28Aug/SortMatrixbyDiagonal.cpp
//Time Complexity:O(n^2)
//Space Complexity:O(n)
class Solution {
public:
    vector<vector<int>> sortMatrix(vector<vector<int>>& grid) {
        int n=grid.size();
        // int n=grid.size();
        unordered_map<int,vector<int>>mp;
        vector<vector<int>>mat(n,vector<int>(n,0));

        for(int i=0;i<n;i++){
            for(int j=0;j<n;j++){
                mp[i-j].push_back(grid[i][j]);
                
            }
        }
        for(auto &it:mp){
            if(it.first>=0){
                sort(it.second.begin(),it.second.end(),greater<int>());
            }
            else{
                sort(it.second.begin(),it.second.end());
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                mat[i][j] = mp[i - j].front();
                mp[i - j].erase(mp[i - j].begin());
            }
        }
        return mat;

    }
};